pyseir/backtest/backtest.py

```diff
@@ -0,0 +1,432 @@
+    
+    return prediction
+
+
+def run_backtest(fips,
+                 observations,
+                 rolling_window_size=1,
+                 prediction_window_size=7,
+                 max_observation_days_blinded=40,
+                 error_types=['nrmse', 'rmse', 'relative_error',
+                              'percentage_abs_error', 'symmetric_abs_error'],
+                 ref_date=datetime(year=2020, month=1, day=1),
+                 n_retries=3):
+    """
+    Run backtest for a given fips. Backtest blinds last n days of observations (
+    new cases, new deaths or hospitalizations) and calculates error of
+    prediction in prediction_window_size days since the last un-blinded
+    observation. Errors can be calculated for ith day of prediction where 1
+    <= i <= prediction_window_size or for the entire prediction time window,
+    depending on type of errors.
+
+    Parameters
+    ----------
+    fips: str
+        FIPS code of the state or county to run backtest.
+    rolling_window_size: int
+        Width of time window to aggregate (average) the observation and
+        prediction to calculate prediction error.
+    prediction_window_size: int
+        Size of time window of prediction to calculate prediction error.
+    max_observation_days_blinded: int
+        Maximum number of days to blind the observations.
+    error_types: str
+        Type of prediction error to calculate, should be included in
+        timeseries_metrics.ErrorType.
+    ref_date: Datetime
+        Reference start date.
+    n_retries: int
+        Number of times to retry model fitting.
+
+    Returns
+    -------
+    backtest_results: pd.DataFrame
+        Contains errors of predictions by days of prediction and type of error.
+        With columns:
+        - observation_type: str, type of observations, should be among: 
+                            'new_cases', 'new_deaths', or 'current_hosp'
+        - error_type: str, type of forecasting errors. 
+        - error: float, value of corresponding type of error
+        - days_of_forecast: int, number of days of forecast since latest
+                            available observation.
+        - observation_end_date: Datetime, last day before blinded observations.
+    historical_predictions: pd.DataFrame
+        Contains predictions given historial observation timeseries.
+        With columns:
+        - dates: Datatime, dates of prediction
+        - new_cases: float, predicted new cases
+        - new_deaths: float, predicted new deaths
+        - curret_hosp: float, predicted hospitalizations
+        - observation_end_date: Datetime, last day before blinded observations.
+        - observation_days_blinded: int, number of days that observations
+                                    are blinded.
+    """
+    tsm = TimeSeriesMetrics()
+    backtest_results = list()
+    historical_predictions = list()
+    for d in range(1, max_observation_days_blinded + 1):
```

Comment:
Not sure how long this takes, but could probably add a `step_size` parameter here to speed things up, especially if we are using more smoothing.